skimage/morphology/_util.py

```diff
@@ -51,35 +54,206 @@ def _validate_connectivity(image_dim, connectivity, offset):
     return c_connectivity, offset
 
 
-def _offsets_to_raveled_neighbors(image_shape, structure, center, order='C'):
+def _offsets_to_raveled_neighbors(image_shape, selem, center, order='C'):
     """Compute offsets to a samples neighbors if the image would be raveled.
 
     Parameters
     ----------
     image_shape : tuple
         The shape of the image for which the offsets are computed.
-    structure : ndarray
+    selem : ndarray
         A structuring element determining the neighborhood expressed as an
         n-D array of 1's and 0's.
-    center : sequence
-        Tuple of indices specifying the center of `selem`.
+    center : tuple
+        Tuple of indices to the center of `selem`.
+    order : {"C", "F"}, optional
+        Whether the image described by `image_shape` is in row-major (C-style)
+        or column-major (Fortran-style) order.
 
     Returns
     -------
-    offsets : ndarray
+    raveled_offsets : ndarray
         Linear offsets to a samples neighbors in the raveled image, sorted by
-        their Euclidean distance from the center.
+        their distance from the center.
+
+    Notes
+    -----
+    This function will return values even if `image_shape` contains a dimension
+    length that is smaller than `selem`.
 
     Examples
     --------
     >>> _offsets_to_raveled_neighbors((4, 5), np.ones((4, 3)), (1, 1))
     array([-5, -1,  1,  5, -6, -4,  4,  6, 10,  9, 11])
+    >>> _offsets_to_raveled_neighbors((2, 3, 2), np.ones((3, 3, 3)), (1, 1, 1))
+    array([ 2, -6,  1, -1,  6, -2,  3,  8, -3, -4,  7, -5, -7, -8,  5,  4, -9,
+            9])
+    """
+    if not selem.ndim == len(image_shape) == len(center):
+        raise ValueError(
+            "number of dimensions in image_shape, selem and center "
+            "does not match"
+        )
+
+    selem_indices = np.array(np.nonzero(selem)).T
+    offsets = selem_indices - center
+
+    if order == 'F':
+        offsets = offsets[:, ::-1]
+        image_shape = image_shape[::-1]
+    elif not order == 'C':
+        raise ValueError("order was not 'C' or 'F'")
+
+    # Scale offsets in each dimension and sum
+    ravel_factors = image_shape[1:] + (1,)
+    ravel_factors = np.cumprod(ravel_factors[::-1])[::-1]
+    raveled_offsets = (offsets * ravel_factors).sum(axis=1)
+
+    # Sort by distance
+    distances = np.abs(offsets).sum(axis=1)
+    raveled_offsets = raveled_offsets[np.argsort(distances)]
+
+    # If any dimension in image_shape is smaller than selem.shape
+    # duplicates might occur, remove them
+    if any(x < y for x, y in zip(image_shape, selem.shape)):
```

Comment:
couldn't you just add this part to the existing function? I'm curious to know why you modified the rest of the function, since I find it harder to read than the original one. It might be only a matter of taste of course...

Reply:
Short answer: No. :wink: I probably should have done a better job highlighting the problem and what I did about it. Consider the following scenario:

I have an `image` of shape `(10, 2)` and want to compute the neighborhood offsets for a structuring element with the shape `(3, 3)` with center at `(1, 1)`. This is a valid use case we need to anticipate because we might want the algorithm be able to step from `image[i, 0]` to `image[i, 1 ]` *and* vice versa. For that we need neighbors on both sides of the center and thus a dimension length of 3.

Now the old implementation would fail because [np.ravel_multi_index](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ravel_multi_index.html) doesn't support this use case:
```python
import numpy as np
selem = np.ones((3, 3), dtype=bool)
np.ravel_multi_index(np.nonzero(selem), dims=(10, 2))
# Raises Exception: ...
#     np.ravel_multi_index(np.nonzero(selem), image.shape)
#  File "<__array_function__ internals>", line 6, in ravel_multi_index
# ValueError: invalid entry in coordinates array
np.ravel_multi_index(np.nonzero(selem), dims=(10, 3))
# Out: array([0, 1, 2, 3, 4, 5, 6, 7, 8])
```
So the previous implementation wasn't able to deal with this case because it relied on that function to compute the offsets. Unfortunately the `mode` argument is not helpful here because wrapping or clipping is not what we want. I tried adjusting the indices by the center index beforehand but again `ravel_multi_index` doesn't handle negative indices the way we want to.

But the new algorithm will lead to duplicate neighbors. I find it really hard to think about this but I'll try to explain with an example. We have the structuring element with the shape `(3, 3)` all ones and the center `(1, 1)` and a C-ordered image with the shape `(6, 2)`. After adjusting for the center we get the raveled offsets (representation with index pairs ordered by dimension):
```python
[[(-1, -1), (-1,  0), (-1,  1)],
 [( 0, -1), ( 0,  0), ( 0,  1)],
 [( 1, -1), ( 1,  0), ( 1,  1)]]
```
The second value in a pair denotes the step in the last dimension which is contiguous (-> ravel factor 1). But because samples in the first dimension don't border each other we need to address this by multiplying with the length of the previous dimension (-> ravel factor 2). For `(0, 1)` and `(1, -1)` this results `(0, 1)` and `[2, -1]` which both resolve (by summing) to the raveled index 1. Considering the array
```python
[[ 0,  1],
 [ 2,  3],
 [ 4,  5],
 [ 6,  7],
 [ 8,  9],
 [10, 11]]
```
this means that the pixel to the right of `6` is `7` and to the lower-right is `7` as well! This "memory-wise wrapping" when crossing image boundaries might be counter-intuitive to simply wrapping around in one dimension. Though, for the functions using this result it's not relevant because they always check that the image edge is not crossed while iterating the neighborhood.

It's also interesting to switch the image shape `(2, 7)` around (or use F-order) in which case we get no duplicates value wise `[-7, -6, -5, -1, 0, 1, 5,  6, 7]`. However we get the same wrapping behavior.

Okay, I hope that helped. :sweat_smile: 

Though, I wholeheartedly agree that the old implementation is simpler, easier to read and probably to maintain but I saw no way around it.

Reply:
Very interesting! Thanks for the exposé, @lagru! Stupid edge cases ruining our otherwise elegant lives... =P I'll try to do a proper review of this PR in the next couple of days... Thanks!

Reply:
See also https://github.com/numpy/numpy/issues/15475.